state/src/item/metadata.rs

```diff
@@ -124,25 +134,33 @@ impl CacheableItem for Metadata {
 
 const PREFIX: u8 = super::METADATA_PREFIX;
 
+const INITIAL_LEN: usize = 4;
+const TERM_LEN: usize = 2;
+const PARAMS_LEN: usize = 2;
+const TERM_PARAMS_LEN: usize = 1;
+
 impl Encodable for Metadata {
     fn rlp_append(&self, s: &mut RlpStream) {
-        const INITIAL_LEN: usize = 4;
-        const TERM_LEN: usize = 2;
-        const PARAMS_LEN: usize = 2;
-        let mut len = INITIAL_LEN;
-
         let term_changed = self.term != Default::default();
+        let params_changed = self.seq != 0;
+        let term_params_changed = self.term_params.is_some();
+
+        let mut len = INITIAL_LEN;
         if term_changed {
             len += TERM_LEN;
         }
-
-        let params_changed = self.seq != 0;
         if params_changed {
             if !term_changed {
                 len += TERM_LEN;
             }
             len += PARAMS_LEN;
         }
+        if term_params_changed {
+            if !params_changed {
+                panic!("Term params only can be changed if params changed");
+            }
+            len += TERM_PARAMS_LEN;
+        }
         s.begin_list(len)
```

Comment:
How about using `begin_unbounded_list` instead?
You don't have to calculate the length if you use it.

Reply:
I used `begin_list` as a replacement for the `assert`s.

Reply:
I think it would be far simpler if we use `begin_unbounded_list`, but it seems your intention has a point too.
Let's leave it as is.